kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -436,6 +440,124 @@ private static Map<Long, StructField> fieldsById(StructType schema) {
     return columnIdToField;
   }
 
+  private static void validateNestedIds(
+      SchemaChanges schemaChanges,
+      Map<Long, StructField> currentFieldsById,
+      Map<Long, StructField> updatedFieldsById) {
+    for (Tuple2<StructField, StructField> updatedField : schemaChanges.updatedFields()) {
+      validateNestedIdsExist(updatedField._2);
+    }
+
+    for (StructField addedField : schemaChanges.addedFields()) {
+      validateNestedIdsExist(addedField);
+    }
+
+    validateNestedIdConsistency(currentFieldsById, updatedFieldsById);
+  }
+
+  private static void validateNestedIdsExist(StructField field) {
+    if (field.getDataType() instanceof MapType || field.getDataType() instanceof ArrayType) {
+      checkArgument(
+          hasNestedColumnIds(field),
+          "Field %s is missing expected nested column IDs",
+          field.getName());
+      if (field.getDataType() instanceof MapType) {
+        checkArgument(
+            getNestedFieldIds(field).size() >= 2,
+            "Map field %s needs at least 2 nested IDs, one for the key and one for the value",
+            field.getName());
+      }
+    }
+  }
+
+  /* Mapping of field ID to nested IDs for that field */
+  private static Map<Long, List<NestedIdEntry>> columnIdToNestedIds(
+      Map<Long, StructField> currentFieldIdToField) {
+    Map<Long, List<NestedIdEntry>> nestedIdEntries = new HashMap<>();
+    for (Map.Entry<Long, StructField> entry : currentFieldIdToField.entrySet()) {
+      if (hasNestedColumnIds(entry.getValue())) {
+        nestedIdEntries.put(entry.getKey(), getNestedFieldIds(entry.getValue()));
+      }
+    }
+
+    return nestedIdEntries;
+  }
+
+  /**
+   * Validate that any nested IDs that were in the current field are consistent with the ones in the
```

Comment:
what does "consistent" mean? equal to? unchanged? maybe tweak that verbiage?

Reply:
Unchanged is a better indicator of what I mean here